### PR TITLE
dev-scheme/racket: miscellaneous fixes

### PR DESCRIPTION
**Pull request by xgqt**
Important:
  - some dependencies are required only for GRacket (USE="X")
  - unset PLTUSERHOME - when set breaks src_install

Misc:
  - reset XDG environment
  - econf array

With X:
  - create racket icon
  - create missing launchers for gracket & plt-games
  - regenerate XDG databases

Package-Manager: Portage-3.0.18, Repoman-3.0.2
Signed-off-by: Maciej Barć <xgqt@riseup.net>

### CODE REVIEW

**Comment by xgqt** [comment]
CC @thesamesam

**Comment by xgqt** [comment]
+ rev bump?

**Comment by xgqt** [comment]
old changes had conflicts with current HEAD so I thought I will add the newest versions 8.1 & 8.2

The changes to racket ebuild are described here: https://src_prepare.gitlab.io/racket/racket-overlay/racket-overlay-install.html (in section "Racket’s ebuild changes")

cc @thesamesam

File: dev-scheme/racket/racket-8.2.ebuild [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+		x11-misc/shared-mime-info
+	)
+	readline? ( dev-libs/libedit )
+"
+RDEPEND="${DEPEND}"
+
+# BC library is being stripped unconditionally
+QA_PRESTRIPPED="/usr/lib64/libracket3m-${PV}.so"
+
+post_X_update() {
```
Comment (thesamesam) [inline dev-scheme/racket/racket-8.2.ebuild:48]: Actually, let's just do:
```
use X && xdg_pkg_postinst
```
and so on in pkg_postinst (and its friend) and drop this?

(Inherit xdg.eclass instead).

File: dev-scheme/racket/racket-8.2.ebuild [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+		x11-libs/gtk+:3[X]
+		x11-libs/pango[X]
+		x11-misc/shared-mime-info
+	)
+	readline? ( dev-libs/libedit )
+"
+RDEPEND="${DEPEND}"
+
+# BC library is being stripped unconditionally
+QA_PRESTRIPPED="/usr/lib64/libracket3m-${PV}.so"
```
Comment (thesamesam) [inline dev-scheme/racket/racket-8.2.ebuild:46]:
```suggestion
QA_PRESTRIPPED="usr/lib*/libracket3m-${PV}.so"
```

Paths aren't absolute in theory, IIRC.

File: dev-scheme/racket/racket-8.2.ebuild [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+	xdg_environment_reset
+
+	default
+
+	# Remove bundled libffi
+	rm -r ./bc/foreign/libffi || die "failed to remove bundled libffi"
+}
+
+src_configure() {
+	# According to vapier, we should use the bundled libtool
```
Comment (thesamesam) [inline dev-scheme/racket/racket-8.2.ebuild:67]: heh

File: dev-scheme/racket/racket-8.2.ebuild [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+	if use doc; then
+		docompress -x "/usr/share/doc/${PF}"
+	fi
+
+	# Remove unneeded "*.a" and "*.la"
+	find "${ED}" \( -name "*.a" -o -name "*.la" \) -delete || die
+
+	# Create missing desktop files and icon
+	if use X; then
+		newicon "${D}/usr/share/racket/drracket-exe-icon.png" "racket.png"
```
Comment (thesamesam) [inline dev-scheme/racket/racket-8.2.ebuild:139]:
```suggestion
		newicon "${ED}/usr/share/racket/drracket-exe-icon.png" "racket.png"
```

File: dev-scheme/racket/racket-8.2.ebuild [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+		/usr/share/racket/links.rktd
+		/usr/share/racket/pkgs/pkgs.rktd
+	)
+	# If we are merging the same version (not revision!)
+	# check if pkg database files exist and do not overwrite them
+	if has_version "~${CATEGORY}/${PN}-${PV}"; then
+		echo "We are installing the same version: ${PV}"
+		local rktd
+		for rktd in "${pkgdb[@]}"; do
+			if [ -f "${EROOT}/${rktd}" ]; then
```
Comment (thesamesam) [inline dev-scheme/racket/racket-8.2.ebuild:158]: Bash test [please](https://devmanual.gentoo.org/tools-reference/bash/#single-versus-double-brackets-in-bash)!

File: dev-scheme/racket/racket-8.2.ebuild [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+				einfo "Keeping old file: ${rktd}"
+				cp "${EROOT}/${rktd}" "${D}/${rktd}" || die
```
Comment (thesamesam) [inline dev-scheme/racket/racket-8.2.ebuild:160]:
```suggestion
				cp "${EROOT}"/${rktd} "${ED}"/${rktd} || die
```

File: dev-scheme/racket/racket-8.2.ebuild [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+			pax-mark m "${D}/usr/bin/${f}"
+		done
+
+		pax-mark m "${D}/usr/$(get_libdir)/racket/starter"
+		use X && pax-mark m "${D}/usr/$(get_libdir)/racket/gracket"
+	fi
+
+	# raco needs decompressed files for packages doc installation bug 662424
+	if use doc; then
+		docompress -x "/usr/share/doc/${PF}"
```
Comment (thesamesam) [inline dev-scheme/racket/racket-8.2.ebuild:131]: Nit: avoid overquoting

```suggestion
		docompress -x /usr/share/doc/${PF}
```

File: dev-scheme/racket/racket-8.2.ebuild [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+src_install() {
+	default
+
+	# When the JIT is enabled, the final binaries need to be pax-marked, too,
+	# if you want to actually use them. The src_compile marking get lost
+	# somewhere in the install process.
+	if use jit; then
+		local f
+		for f in mred mzscheme racket; do
+			pax-mark m "${D}/usr/bin/${f}"
```
Comment (thesamesam) [inline dev-scheme/racket/racket-8.2.ebuild:122]: If we know a path isn't prefixed, let's use ${ED}.
```suggestion
			pax-mark m "${ED}"/usr/bin/${f}
```

(Also dropped overquoting, but it's a nit).

File: dev-scheme/racket/racket-8.2.ebuild [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+		done
+
+		pax-mark m "${D}/usr/$(get_libdir)/racket/starter"
```
Comment (thesamesam) [inline dev-scheme/racket/racket-8.2.ebuild:125]:
```suggestion
		pax-mark m "${ED}/usr/$(get_libdir)/racket/starter"
```

File: dev-scheme/racket/racket-8.2.ebuild [hunk 10]
```diff
@@ -0,0 +1,172 @@
...
+		use X && pax-mark m "${D}/usr/$(get_libdir)/racket/gracket"
```
Comment (thesamesam) [inline dev-scheme/racket/racket-8.2.ebuild:126]:
```suggestion
		use X && pax-mark m "${ED}/usr/$(get_libdir)/racket/gracket"
```

**Review by thesamesam — CHANGES_REQUESTED** [review]

File: dev-scheme/racket/racket-8.2.ebuild (hunk 1 above)
Comment (xgqt) [reply dev-scheme/racket/racket-8.2.ebuild:48]: > xdg.eclass 

then we get unwanted deps: dev-util/desktop-file-utils & x11-misc/shared-mime-info

File: dev-scheme/racket/racket-8.2.ebuild (hunk 3 above)
Comment (xgqt) [reply dev-scheme/racket/racket-8.2.ebuild:67]: ?

File: dev-scheme/racket/racket-8.2.ebuild (hunk 1 above)
Comment (ionenwks) [reply dev-scheme/racket/racket-8.2.ebuild:48]: I do wish the second point in this would go through :/
https://archives.gentoo.org/gentoo-dev/message/44462535c5d5e5194994ff5a24d0fb6c

But guess the push for IDEPEND is distracting from this option.

On a side-note, with xdg.eclass `use X &&` wouldn't be needed, it's designed to do nothing if there's no .desktop and no icons needing cache updates.

Comment (thesamesam) [reply dev-scheme/racket/racket-8.2.ebuild:48]: Oh, of course. But yes, the dep :|

Comment (xgqt) [reply dev-scheme/racket/racket-8.2.ebuild:48]: > I do wish the second point in this would go through :/
> https://archives.gentoo.org/gentoo-dev/message/44462535c5d5e5194994ff5a24d0fb6c

this would be nice


> On a side-note, with xdg.eclass `use X &&` wouldn't be needed, it's designed to do nothing if there's no .desktop and no icons needing cache updates.

true! good to know

**Comment by xgqt** [comment]
in current HEAD it works with EAPI 8

**Comment by xgqt** [comment]
issue with "`starter`" ignoring `LDFLAGS` will be fixed in next release, it not a big deal though
https://github.com/racket/racket/pull/3945

**Comment by xgqt** [comment]
let's bump EAPI after merge maybe

**Comment by orlitzky** [comment]
I added that crazy pax-mark stuff in a710d5349, but these days it's "obsolete" and you can remove it if you want. PaX was cool and it still exists, but there's no freely available implementation anymore, so the gymnastics are likely ineffective and no one can test them.

Otherwise, unless @Amynka objects, I think we should get out of his way...

**Comment by xgqt** [comment]
> I added that crazy pax-mark stuff in [a710d53](https://github.com/gentoo/gentoo/commit/a710d534973360bae165a2d45cde301dd9956c70), but these days it's "obsolete" and you can remove it if you want. PaX was cool and it still exists, but there's no freely available implementation anymore, so the gymnastics are likely ineffective and no one can test them.
> 
> Otherwise, unless @Amynka objects, I think we should get out of his way...

June 2021: 1 commit

So... another 3 months of waiting?

File: dev-scheme/racket/racket-8.1.ebuild [hunk 11]
```diff
@@ -0,0 +1,181 @@
...
+	)
+	readline? ( dev-libs/libedit )
+"
+RDEPEND="${DEPEND}"
+
+# "mred" and "mzscheme" are binaries generated by Racket (not CC)
+QA_FLAGS_IGNORED="usr/bin/mred usr/bin/mzscheme"
+
+# BC library is being stripped unconditionally
+QA_PRESTRIPPED="usr/lib*/libracket3m-${PV}.so"
```
Comment (thesamesam) [inline dev-scheme/racket/racket-8.1.ebuild:49]: Why don't we prevent stripping it?

**Review by thesamesam — CHANGES_REQUESTED** [review]

File: dev-scheme/racket/racket-8.1.ebuild (hunk 11 above)
Comment (xgqt) [reply dev-scheme/racket/racket-8.1.ebuild:49]: Thought this would work, but not, gotta keep looking...

```patch
index 3393e0c55e..c77b28a4cb 100644
--- a/bc/Makefile.in
+++ b/bc/Makefile.in
@@ -526,7 +526,6 @@ unix-install-3m:
 unix-install-libs-3m:
 	cd ..; $(ICP_LIB) bc/libracket3m.@LIBSFX@ "$(DESTDIR)$(libdir)/libracket3m.@LIBSFX@"
 	cd ..; $(ICP_STATIC_LIB) bc/rktio/librktio.@LIBSFX@ "$(DESTDIR)$(libdir)/librktio.@LIBSFX@"
-	cd ..; $(STRIP_LIB_DEBUG) "$(DESTDIR)$(libdir)/libracket3m.@LIBSFX@"
 
 unix-no-install-libs-3m:
 	$(NOOP)
```

File: dev-scheme/racket/files/Makefile-nostrip.patch [hunk 12]
```diff
@@ -0,0 +1,23 @@
...
+-	cd ..; $(STRIP_LIB_DEBUG) "$(DESTDIR)$(libdir)/libracket.@LIBSFX@"
+ 
+ unix-no-install-libs-cgc:
+ 	$(NOOP)
+@@ -524,9 +522,8 @@ unix-install-3m:
+ 	$(RESTORE_SIGNATURE) "$(DESTDIR)$(bindir)/racket@MMM_INSTALLED@"
+ 
+ unix-install-libs-3m:
+-	cd ..; $(ICP_LIB) bc/libracket3m.@LIBSFX@ "$(DESTDIR)$(libdir)/libracket3m.@LIBSFX@"
++	cd ..; libtool --mode=install install bc/libracket3m.@LIBSFX@ "$(DESTDIR)$(libdir)/libracket3m.@LIBSFX@"
```
Comment (thesamesam) [inline dev-scheme/racket/files/Makefile-nostrip.patch:18]: We probably don't want to call libtool directly, but instead use the "bundled" / generated one. That might be $(ICP_LIB) in this case (not sure).

Is there a reason you needed to drop it?

Comment (xgqt) [reply dev-scheme/racket/files/Makefile-nostrip.patch:18]: > We probably don't want to call libtool directly, but instead use the "bundled" / generated one. That might be $(ICP_LIB) in this case (not sure).
> 
> Is there a reason you needed to call directly?

ICP_LIB was a call to libtool that included command to strip the library

in logs I saw this
```log
/var/tmp/portage/dev-scheme/racket-8.2/work/racket-8.2/src/bc/rktio/../lt/libtool --mode=link --tag=CC x86_64-pc-li
nux-gnu-gcc -o  librktio.la rktio_fs.lo rktio_fd.lo rktio_file.lo rktio_poll_set.lo rktio_ltps.lo rktio_sleep.lo rk
tio_network.lo rktio_pipe.lo rktio_process.lo rktio_signal.lo rktio_envvars.lo rktio_fs_change.lo rktio_flock.lo rk
tio_shellex.lo rktio_time.lo rktio_cpu.lo rktio_syslog.lo rktio_convert.lo rktio_sha1.lo rktio_sha2.lo rktio_dll.lo
 rktio_error.lo rktio_hash.lo rktio_wide.lo rktio_console.lo rktio_main.lo

 ...

 libtool --mode=compile --tag=CC x86_64-pc-linux-gnu-gcc -march=znver1 -O3 -falign-functions=32 -fno-common -fstack-protector-strong -pipe -fdiagnostics-color=always -frecord-gcc-switches  -Wall   -I./.. -I./../include -I./../../version -I./../../rktio -I../rktio -pthread -I/usr/lib64/libffi/include   -DUSE_SENORA_GC  -DMZ_USES_SHARED_LIB -c ./salloc.c -o salloc.lo
 ```
 which means it sometimes calls bundled libtool and sometimes it doesn't
 
 
 In case of libracket3m it wasn't calling the bundled one:
 ```log
 make unix-install-libs-3m
make[5]: Entering directory '/var/tmp/portage/dev-scheme/racket-8.2/work/racket-8.2/src/bc'
cd ..; libtool --mode=install install -s bc/libracket3m.la "/var/tmp/portage/dev-scheme/racket-8.2/image/usr/lib64/
libracket3m.la"
libtool: install: install bc/.libs/libracket3m-8.2.so /var/tmp/portage/dev-scheme/racket-8.2/image/usr/lib64/librac
ket3m-8.2.so
libtool: install: x86_64-pc-linux-gnu-strip --strip-unneeded /var/tmp/portage/dev-scheme/racket-8.2/image/usr/lib64
/libracket3m-8.2.so
libtool: install: (cd /var/tmp/portage/dev-scheme/racket-8.2/image/usr/lib64 && { ln -s -f libracket3m-8.2.so libra
cket3m.so || { rm -f libracket3m.so && ln -s libracket3m-8.2.so libracket3m.so; }; })
libtool: install: install bc/.libs/libracket3m.lai /var/tmp/portage/dev-scheme/racket-8.2/image/usr/lib64/libracket
3m.la
libtool: install: install bc/.libs/libracket3m.a /var/tmp/portage/dev-scheme/racket-8.2/image/usr/lib64/libracket3m
.a
libtool: install: x86_64-pc-linux-gnu-strip --strip-debug /var/tmp/portage/dev-scheme/racket-8.2/image/usr/lib64/li
bracket3m.a
libtool: install: chmod 644 /var/tmp/portage/dev-scheme/racket-8.2/image/usr/lib64/libracket3m.a
libtool: install: x86_64-pc-linux-gnu-ranlib /var/tmp/portage/dev-scheme/racket-8.2/image/usr/lib64/libracket3m.a
libtool: warning: remember to run 'libtool --finish /usr/lib64'
```

Comment (thesamesam) [reply dev-scheme/racket/files/Makefile-nostrip.patch:18]: That's a bit annoying, but it's not your job to necessarily fix this.

That said, I'd prefer to keep $(ICP_LIB) and just drop whereever the strip is coming from, but if we can't do that, so be it.

Comment (xgqt) [reply dev-scheme/racket/files/Makefile-nostrip.patch:18]: > That said, I'd prefer to keep $(ICP_LIB) and just drop whereever the strip is coming from, but if we can't do that, so be it.

Those are the changes it took to fix it the way you want: https://github.com/racket/racket/commit/92d202060201f569fafc348c4f64a2965b55287f

I opted to have a simpler patch.

Comment (thesamesam) [reply dev-scheme/racket/files/Makefile-nostrip.patch:18]: Works for me for now! Thank you!

Comment (Amynka) [reply dev-scheme/racket/files/Makefile-nostrip.patch:18]: Looks good..

File: dev-scheme/racket/files/Makefile-nostrip.patch [hunk 13]
```diff
@@ -0,0 +1,23 @@
+index 3393e0c55e..6a1c3e150b 100644
+--- a/bc/Makefile.in
++++ b/bc/Makefile.in
+@@ -505,8 +505,6 @@ unix-install-libs-cgc:
+ 	cd ..; $(ICP_LIB) bc/libmzgc.@LIBSFX@ "$(DESTDIR)$(libdir)/libmzgc.@LIBSFX@"
+ 	cd ..; $(ICP_LIB) bc/libracket.@LIBSFX@ "$(DESTDIR)$(libdir)/libracket.@LIBSFX@"
+ 	cd ..; $(ICP_STATIC_LIB) bc/rktio/librktio.@LIBSFX@ "$(DESTDIR)$(libdir)/librktio.@LIBSFX@"
+-	cd ..; $(STRIP_LIB_DEBUG) "$(DESTDIR)$(libdir)/libmzgc.@LIBSFX@"
+-	cd ..; $(STRIP_LIB_DEBUG) "$(DESTDIR)$(libdir)/libracket.@LIBSFX@"
```
Comment (ulm) [inline dev-scheme/racket/files/Makefile-nostrip.patch:9]: I wonder about this patch. AFAICS the `--disable-strip` configure option already causes `STRIP_LIB_DEBUG` to be defined as a no-op (namely `:`), so why is this needed in addition?

File: dev-scheme/racket/racket-8.1.ebuild [hunk 14]
```diff
@@ -0,0 +1,184 @@
...
+	# Remove bundled libffi
+	rm -r ./bc/foreign/libffi || die "failed to remove bundled libffi"
+}
+
+src_configure() {
+	# According to vapier, we should use the bundled libtool
+	#   such that we don't preclude cross-compile.
+	#   Thus don't use --enable-lt=/usr/bin/libtool
+	# --enable-bc builds Racket w/o chez backend
+	local myconf=(
```
Comment (ulm) [inline dev-scheme/racket/racket-8.1.ebuild:81]: No need for an intermediate variable here. These options can be passed to `econf` directly.

File: dev-scheme/racket/racket-8.1.ebuild [hunk 15]
```diff
@@ -0,0 +1,184 @@
...
+	# When the JIT is enabled, a few binaries need to be pax-marked
+	# on hardened systems (bug 613634). The trick is to pax-mark
+	# them before they're used later in the build system. The
+	# following order for racketcgc and racket3m was determined by
+	# digging through the Makefile in src/racket to find out which
+	# targets would build those binaries but not use them.
+	if use jit && ! use chez; then
+		pushd ./bc  || die
+		emake cgc-core
+		pax-mark m .libs/racketcgc
```
Comment (ulm) [inline dev-scheme/racket/racket-8.1.ebuild:110]: We no longer have packages for a PaX kernel, so IMHO it is no longer a supported configuration. So you may consider dropping all the `pax-mark` logic at some point. It would considerably simplify the ebuild.

Of course, this isn't a blocker for this PR.

File: dev-scheme/racket/files/Makefile-nostrip.patch (hunk 13 above)
Comment (xgqt) [reply dev-scheme/racket/files/Makefile-nostrip.patch:9]: true, I removed this while searching for solution, forgot to remove from patch

File: dev-scheme/racket/racket-8.1.ebuild (hunk 14 above)
Comment (xgqt) [reply dev-scheme/racket/racket-8.1.ebuild:81]: sure they can, but it looks awful i my hones opinion with larger configurations, where we need to escape newline `\`

File: dev-scheme/racket/racket-8.1.ebuild [hunk 16]
```diff
@@ -0,0 +1,153 @@
...
+	# pkg database files
+	local pkgdb=(
+		/usr/share/racket/info-cache.rktd
+		/usr/share/racket/links.rktd
+		/usr/share/racket/pkgs/pkgs.rktd
+	)
+	# If we are merging the same version (NOT revision!)
+	# check if pkg database files exist and do not overwrite them
+	local rel
+	for pvr in ${REPLACING_VERSIONS}; do
```
Comment (ulm) [inline dev-scheme/racket/racket-8.1.ebuild:129]: `pvr` should be declared local too.

**Review by ulm — COMMENTED** [review]
Rest looks good.

**Review by ulm — APPROVED** [review]
